File: src/snippets/bot-protection/reference/bun/IdentifiedBots.ts
import arcjet, { detectBot } from "@arcjet/bun";
import { env } from "bun";

const aj = arcjet({
  key: env.ARCJET_KEY!, // Get your site key from https://app.arcjet.com
  rules: [
    detectBot({
      mode: "LIVE", // will block requests. Use "DRY_RUN" to log only
      allow: [], // "allow none" will block all detected bots
    }),
  ],
});

export default {
  port: 3000,
  fetch: aj.handler(async (req) => {
    const decision = await aj.protect(req);

    for (const { reason } of decision.results) {
      if (reason.isBot()) {
        console.log("detected + allowed bots", reason.allowed);
        console.log("detected + denied bots", reason.denied);

        // Arcjet Pro plan verifies the authenticity of common bots using IP data
        // https://docs.arcjet.com/bot-protection/reference#bot-verification
        if (reason.isSpoofed()) {
          console.log("spoofed bot", reason.spoofed);
        }

        if (reason.isVerified()) {
          console.log("verified bot", reason.verified);
        }
      }
    }

    if (decision.isDenied()) {
      return new Response("Forbidden", { status: 403 });
    }

    return new Response("Hello world");
  }),
};

File: src/snippets/bot-protection/reference/python-flask/IdentifiedBots.py
import logging
import os

from arcjet import Mode, arcjet_sync, detect_bot, is_spoofed_bot
from flask import Flask, jsonify, request

app = Flask(__name__)

logger = logging.getLogger(__name__)

aj = arcjet_sync(
    key=os.environ["ARCJET_KEY"],  # Get your site key from https://app.arcjet.com
    rules=[
        detect_bot(
            mode=Mode.LIVE,  # Blocks requests. Use Mode.DRY_RUN to log only
            allow=[],  # "allow none" will block all detected bots
        ),
    ],
)


@app.get("/")
def index():
    decision = aj.protect(request)

    for result in decision.results:
        if result.reason_v2.type == "BOT":
            logger.info("detected + allowed bots %s", result.reason_v2.allowed)
            logger.info("detected + denied bots %s", result.reason_v2.denied)

            # Arcjet Pro plan verifies the authenticity of common bots using
            # IP data — see https://docs.arcjet.com/bot-protection/reference#bot-verification
            if result.reason_v2.spoofed:
                logger.info("spoofed bot detected")

            if result.reason_v2.verified:
                logger.info("verified bot detected")

    if decision.is_denied():
        return jsonify(error="Forbidden"), 403

    # The is_spoofed_bot() helper is a convenience wrapper around the above
    if any(is_spoofed_bot(r) for r in decision.results):
        return jsonify(error="Forbidden"), 403

    return jsonify(message="Hello world")
